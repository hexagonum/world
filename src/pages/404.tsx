import { NextPage } from 'next';

export const NotFoundPage: NextPage = () => {
  return (
    <div className="bg-gray-900 text-gray-100 h-screen">
      <div className="flex items-center justify-center h-full">
        <div className="text-center">
          <div className="text-9xl font-bold">404</div>
          <div className="text-3xl uppercase">Page Not Found</div>
        </div>
      </div>
    </div>
  );
};

export default NotFoundPage;
