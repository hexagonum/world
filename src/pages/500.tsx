import ErrorTemplate from '@weather/templates/ErrorTemplate';
import { NextPage } from 'next';

export const InternalServerErrorPage: NextPage = () => <ErrorTemplate code={500} message="Internal Server Error" />;

export default InternalServerErrorPage;
