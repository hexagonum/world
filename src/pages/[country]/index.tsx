import { Badge, Card, CardBody, Divider, TableContainer, Table, Tbody, Tr, Td, TableCaption } from '@chakra-ui/react';
import { Container } from '@weather/components/Container';
import { Weather } from '@weather/components/Weather';
import { cities } from '@weather/data/cities';
import trendsByCountries from '@weather/data/trends.json';
import isoAlpha3Codes from '@weather/data/iso-alpha-3-codes.json';
import unitedNationMembers from '@weather/data/united-nation-members.json';
import { Layout } from '@weather/layout';
import { City } from '@weather/types';
import { NextPage } from 'next';
import Image from 'next/image';
import Link from 'next/link';
import { useRouter } from 'next/router';

const CountryPage: NextPage = () => {
  const { query } = useRouter();
  const countryCode: string = query.country?.toString() || '';
  const country = unitedNationMembers.find(({ cca3 }) => cca3.toLowerCase() === countryCode.toLowerCase()) || {
    name: { common: '', official: '' },
    flags: { svg: '' },
    cca2: '',
    cca3: '',
    ccn3: '',
    cioc: '',
    fifa: '',
    region: '',
    subregion: '',
    latlng: [0, 0],
    area: 0,
    population: 0,
    maps: { googleMaps: '', openStreetMaps: '' },
    idd: { root: '', suffixes: [] },
    capital: [],
    languages: {},
    borders: [],
  };
  const citiesByCountry = cities.filter(
    ({ country: cityCountry }) => country.name.common.toLowerCase() === cityCountry.toLowerCase()
  );
  const trendsByCountry: { trends: string[] } = trendsByCountries.find(
    ({ country: trendCountry }) => country.name.common.toLowerCase() === trendCountry.toLowerCase()
  ) || { trends: [] };
  const { trends = [] } = trendsByCountry;

  return (
    <Layout>
      <Container>
        <div className="p-8">
          <div className="flex flex-col gap-4 md:gap-8">
            <div className="flex flex-col gap-4 md:gap-8">
              <div className="flex items-center justify-between">
                <h1 className="capitalize text-4xl font-bold">{country.name.common}</h1>
                <Image
                  src={country.flags.svg}
                  alt={country.name.common}
                  width={64}
                  height={32}
                  className="border border-solid rounded"
                />
              </div>
              <TableContainer className="border rounded shadow">
                <Table>
                  <Tbody>
                    <Tr>
                      <Td>Capital</Td>
                      <Td isNumeric>{country.capital.join(', ')}</Td>
                    </Tr>
                    <Tr>
                      <Td>Official</Td>
                      <Td isNumeric>{country.name.official}</Td>
                    </Tr>
                    <Tr>
                      <Td>Languages ({Object.keys(country.languages).length})</Td>
                      <Td>
                        <div className="flex flex-wrap justify-end gap-2">
                          {Object.keys(country.languages).map((key: string) => {
                            const value: string = (country.languages as Record<string, string>)[key] || '';
                            return (
                              <Badge key={key} colorScheme="teal" fontSize="sm">
                                {value} ({key})
                              </Badge>
                            );
                          })}
                        </div>
                      </Td>
                    </Tr>
                  </Tbody>
                </Table>
              </TableContainer>
            </div>
            <Divider className="border-gray-300" />
            <div className="flex flex-col gap-4 md:gap-8">
              <h2 className="font-semibold text-xl">Codes</h2>
              <TableContainer className="border rounded shadow">
                <Table>
                  <Tbody>
                    <Tr>
                      <Td>
                        <b>FIFA</b>
                      </Td>
                      <Td isNumeric>{country.fifa}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>IOC Code</b>
                      </Td>
                      <Td isNumeric>{country.cioc}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>ISO Alpha-2 Code</b>
                      </Td>
                      <Td isNumeric>{country.cca2}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>ISO Alpha-3 Code</b>
                      </Td>
                      <Td isNumeric>{country.cca3}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>ISO Numeric Code</b>
                      </Td>
                      <Td isNumeric>{country.ccn3}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>IDD Prefixes ({country.borders.length})</b>
                      </Td>
                      <Td isNumeric>
                        <div className="flex flex-wrap justify-end gap-2">
                          {country.idd.suffixes.map((suffix: string) => (
                            <Badge
                              key={suffix}
                              colorScheme="teal"
                              fontSize="xs"
                            >{`${country.idd.root}${suffix}`}</Badge>
                          ))}
                        </div>
                      </Td>
                    </Tr>
                  </Tbody>
                </Table>
              </TableContainer>
            </div>
            <Divider className="border-gray-300" />
            <div className="flex flex-col gap-4 md:gap-8">
              <h2 className="font-semibold text-xl">Geography</h2>
              <TableContainer className="border rounded shadow">
                <Table>
                  <Tbody>
                    <Tr>
                      <Td>
                        <b>Region</b>
                      </Td>
                      <Td isNumeric>{country.region}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>Subregion</b>
                      </Td>
                      <Td isNumeric>{country.subregion}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <span className="hidden md:block">
                          <b>Latitude / Longitude</b>
                        </span>
                        <span className="block md:hidden">
                          <b>Lat / Lng</b>
                        </span>
                      </Td>
                      <Td isNumeric>
                        <Link href={country.maps.googleMaps}>
                          <Badge colorScheme="teal" variant="outline" fontSize="sm">
                            <span>
                              {country.latlng[0]}&deg;{country.latlng[0] >= 0 ? 'N' : 'S'}
                            </span>
                            {', '}
                            <span>
                              {country.latlng[1]}&deg;
                              {country.latlng[1] >= 0 ? 'E' : 'W'}
                            </span>
                          </Badge>
                        </Link>
                      </Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>Borders ({country.borders.length})</b>
                      </Td>
                      <Td isNumeric>
                        <div className="flex flex-wrap justify-end gap-2">
                          {country.borders.map((border: string) => {
                            return (
                              <Link key={border} href={`/${border}`}>
                                <Badge colorScheme="teal" fontSize="sm">
                                  {(isoAlpha3Codes as Record<string, string>)[border]}
                                </Badge>
                              </Link>
                            );
                          })}
                        </div>
                      </Td>
                    </Tr>
                  </Tbody>
                </Table>
              </TableContainer>
            </div>
            <Divider className="border-gray-300" />
            <div className="flex flex-col gap-4 md:gap-8">
              <h2 className="font-semibold text-xl">Population</h2>
              <TableContainer className="border rounded shadow">
                <Table>
                  <Tbody>
                    <Tr>
                      <Td>
                        <b>Population</b>
                      </Td>
                      <Td isNumeric>{country.population.toLocaleString()}</Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>Area</b>
                      </Td>
                      <Td isNumeric>
                        {country.area.toLocaleString()} km<sup>2</sup>
                      </Td>
                    </Tr>
                    <Tr>
                      <Td>
                        <b>Density</b>
                      </Td>
                      <Td isNumeric>
                        {(country.population / country.area).toFixed(2)} people/km<sup>2</sup>
                      </Td>
                    </Tr>
                  </Tbody>
                </Table>
              </TableContainer>
            </div>
            <Divider className="border-gray-300" />
            {citiesByCountry.length > 0 ? (
              <div className="flex flex-col gap-4 md:gap-8">
                <h2 className="font-semibold text-xl">Weather ({citiesByCountry.length})</h2>
                <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4 md:gap-8">
                  {citiesByCountry.map(({ id, name, latitude, longitude, timezone }: City) => (
                    <Weather key={id} city={name} latitude={latitude} longitude={longitude} timezone={timezone} />
                  ))}
                </div>
              </div>
            ) : (
              <></>
            )}
            <Divider className="border-gray-300" />
            {trends.length > 0 ? (
              <div className="flex flex-col gap-4 md:gap-8">
                <h2 className="font-semibold text-xl">Google Trends ({trends.length})</h2>
                <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-4 md:gap-8">
                  {trends.map((trend: string) => {
                    const query: string = encodeURIComponent(trend);
                    const href: string = `https://www.google.com/search?q=${query}`;
                    return (
                      <Card key={trend} className="border border-gray-200">
                        <CardBody>
                          <Link href={href} target="_blank">
                            <p className="truncate">{trend}</p>
                          </Link>
                        </CardBody>
                      </Card>
                    );
                  })}
                </div>
              </div>
            ) : (
              <></>
            )}
          </div>
        </div>
      </Container>
    </Layout>
  );
};

export default CountryPage;
