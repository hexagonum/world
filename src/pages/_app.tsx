import { ChakraProvider } from '@chakra-ui/react';
import { APP_NAME } from '@weather/configs';
import '@weather/styles/globals.scss';
import type { AppProps } from 'next/app';
import Head from 'next/head';

const App: React.FC<AppProps> = ({ Component, pageProps }: AppProps) => {
  return (
    <>
      <Head>
        <title>{APP_NAME}</title>
      </Head>
      <ChakraProvider>
        <Component {...pageProps} />
      </ChakraProvider>
    </>
  );
};

export default App;
