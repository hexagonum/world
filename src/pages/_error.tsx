import ErrorTemplate from '@world/templates/ErrorTemplate';
import { NextPage } from 'next';

export const ErrorPage: NextPage = () => <ErrorTemplate code={500} message="Error" />;

export default ErrorPage;
