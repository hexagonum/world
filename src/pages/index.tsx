import useFetch from '@weather/hooks/use-fetch';
import { NextPage } from 'next';

const cities: { id: string; name: string; latitude: number; longitude: number }[] = [
  {
    id: 'ho-chi-minh-city',
    name: 'Ho Chi Minh City',
    latitude: 10.82,
    longitude: 106.63,
  },
  {
    id: 'hanoi',
    name: 'Hanoi',
    latitude: 21.02,
    longitude: 105.84,
  },
  {
    id: 'singapore',
    name: 'Singapore',
    latitude: 1.29,
    longitude: 103.85,
  },
  {
    id: 'seoul',
    name: 'Seoul',
    latitude: 37.57,
    longitude: 126.98,
  },
  {
    id: 'melbourne',
    name: 'Melbourne',
    latitude: -37.81,
    longitude: 144.96,
  },
  {
    id: 'dallas',
    name: 'Dallas',
    latitude: 32.78,
    longitude: -96.81,
  },
];

type WeatherResponse = {
  latitude: number;
  longitude: number;
  generationtime_ms: number;
  utc_offset_seconds: number;
  timezone: string;
  timezone_abbreviation: string;
  elevation: number;
  current_weather: {
    temperature: number;
    windspeed: number;
    winddirection: number;
    weathercode: number;
    is_day: number;
    time: string;
  };
};

const weatherCodes: Record<number, string> = {
  0: 'clear sky',
  1: 'mainly clear',
  2: 'partly cloudy',
  3: 'overcast',
  45: 'fog',
  48: 'depositing rime fog',
  51: 'light drizzle',
  53: 'moderate drizzle',
  55: 'dense intensity drizzle',
  56: 'light freezing drizzle',
  57: 'dense intensity freezing drizzle',
  61: 'slight rain',
  63: 'moderate rain',
  65: 'heavy intensity rain',
  66: 'light freezing rain',
  67: 'heavy intensity freezing rain',
  71: 'slight snow fall',
  73: 'moderate snow fall',
  75: 'heavy intensity snow fall',
  77: 'snow grains',
  80: 'slight rain showers',
  81: 'moderate rain showers',
  82: 'violent rain showers',
  85: 'slight snow showers',
  86: 'heavy snow showers',
  95: 'thunderstorm: slight or moderate',
  96: 'slight thunderstorm',
  99: 'heavy hail thunderstorm',
};

type WeatherProps = {
  city: string;
  latitude: number;
  longitude: number;
};

const Weather: React.FC<WeatherProps> = ({ city, latitude, longitude }) => {
  const url = `https://api.open-meteo.com/v1/forecast?latitude=${latitude}&longitude=${longitude}&current_weather=true`;
  const { data, loading, error } = useFetch<WeatherResponse>(url);

  if (loading) {
    return (
      <div className="border border-gray-700 rounded p-4">
        <div className="flex justify-center items-center">Loading</div>
      </div>
    );
  }

  if (error) {
    return (
      <div className="border border-gray-700 rounded p-4">
        <div className="flex justify-center items-center">{error.message}</div>
      </div>
    );
  }

  if (!data) {
    return (
      <div className="border border-gray-700 rounded p-4">
        <div className="flex justify-center items-center">No Data</div>
      </div>
    );
  }

  return (
    <div className="border border-gray-700 rounded p-4">
      <div className="flex justify-between items-center">
        <div className="uppercase">
          <p className="font-bold">{city}</p>
          <p className="text-xs">
            {Math.abs(data.latitude).toFixed(2)}&deg;{data.latitude >= 0 ? 'N' : 'S'} -{' '}
            {Math.abs(data.longitude).toFixed(2)}&deg;{data.longitude >= 0 ? 'E' : 'W'}
          </p>
        </div>
        <div className="text-right">
          <p className="font-bold">{data.current_weather.temperature}&deg;C</p>
          <p className="capitalize text-xs">{weatherCodes[data.current_weather.weathercode] || 'N/A'}</p>
        </div>
      </div>
    </div>
  );
};

export const HomePage: NextPage = () => {
  const year = new Date().getFullYear();

  return (
    <div className="bg-gray-900 text-gray-100 h-screen">
      <div className="flex flex-col h-full">
        <nav className="border-b border-gray-700">
          <div className="container mx-auto px-8 py-4">
            <h1 className="uppercase font-bold">Weather</h1>
          </div>
        </nav>
        <main className="grow overflow-hidden">
          <div className="h-full overflow-auto">
            <div className="container mx-auto p-8">
              <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4 md:gap-8">
                {cities.map((city) => {
                  return <Weather key={city.id} city={city.name} latitude={city.latitude} longitude={city.longitude} />;
                })}
              </div>
            </div>
          </div>
        </main>
        <footer className="border-t border-gray-700">
          <div className="container mx-auto px-8 py-4">
            <p className="uppercase">&copy; {year} Weather</p>
          </div>
        </footer>
      </div>
    </div>
  );
};

export default HomePage;
