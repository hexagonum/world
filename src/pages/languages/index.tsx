import { Table, TableContainer, Tbody, Td, Th, Thead, Tr } from '@chakra-ui/react';
import Container from '@world/components/Container';
import languages from '@world/data/languages/list.json';
import Layout from '@world/layout';
import { NextPage } from 'next';
import Link from 'next/link';

const LanguagesPage: NextPage = () => {
  return (
    <Layout>
      <Container>
        <div className="p-8">
          <TableContainer className="border rounded shadow">
            <Table>
              <Thead>
                <Tr>
                  <Th>Languages ({languages.length})</Th>
                  <Th isNumeric>Countries</Th>
                  <Th isNumeric>Population</Th>
                </Tr>
              </Thead>
              <Tbody>
                {languages.map(({ code = '', name = '', total = 0, population = 0 }) => {
                  return (
                    <Tr key={code}>
                      <Td>
                        <Link href={`/languages/${code}`}>{name}</Link>
                      </Td>
                      <Td isNumeric>{total}</Td>
                      <Td isNumeric>{population.toLocaleString()}</Td>
                    </Tr>
                  );
                })}
              </Tbody>
            </Table>
          </TableContainer>
        </div>
      </Container>
    </Layout>
  );
};

export default LanguagesPage;
