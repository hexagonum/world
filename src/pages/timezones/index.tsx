import { Table, TableContainer, Tbody, Td, Th, Thead, Tr } from '@chakra-ui/react';
import Container from '@world/components/Container';
import timezones from '@world/data/timezones/list.json';
import Layout from '@world/layout';
import { NextPage } from 'next';
import Link from 'next/link';

const TimezonesPage: NextPage = () => {
  return (
    <Layout>
      <Container>
        <div className="p-8">
          <TableContainer className="border rounded shadow">
            <Table>
              <Thead>
                <Tr>
                  <Th>Timezones ({timezones.length})</Th>
                  <Th isNumeric>Offset</Th>
                  <Th isNumeric>Countries</Th>
                </Tr>
              </Thead>
              <Tbody>
                {timezones.map(({ name = '', offset = 0, total = 0 }) => {
                  return (
                    <Tr key={name}>
                      <Td>
                        <Link href={`/timezones/${name}`}>{name}</Link>
                      </Td>
                      <Td isNumeric>{offset}</Td>
                      <Td isNumeric>{total}</Td>
                    </Tr>
                  );
                })}
              </Tbody>
            </Table>
          </TableContainer>
        </div>
      </Container>
    </Layout>
  );
};

export default TimezonesPage;
