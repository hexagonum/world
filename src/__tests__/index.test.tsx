import { render } from '@testing-library/react';
import { HomePage } from '@weather/pages';

describe('HomePage', () => {
  it('to match snapshot', () => {
    const { container } = render(<HomePage />);
    expect(container).toMatchSnapshot();
  });
});
