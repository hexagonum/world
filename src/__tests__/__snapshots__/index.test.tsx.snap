// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage to match snapshot 1`] = `
<div>
  <div
    class="bg-gray-900 text-gray-100 h-screen"
  >
    <div
      class="flex flex-col h-full"
    >
      <nav
        class="border-b border-gray-700"
      >
        <div
          class="container mx-auto px-8 py-4"
        >
          <h1
            class="uppercase font-bold"
          >
            Weather
          </h1>
        </div>
      </nav>
      <main
        class="grow overflow-hidden"
      >
        <div
          class="h-full overflow-auto"
        >
          <div
            class="container mx-auto p-8"
          >
            <div
              class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4 md:gap-8"
            >
              <div
                class="border border-gray-700 rounded p-4"
              >
                <div
                  class="flex justify-center items-center"
                >
                  fetch is not defined
                </div>
              </div>
              <div
                class="border border-gray-700 rounded p-4"
              >
                <div
                  class="flex justify-center items-center"
                >
                  fetch is not defined
                </div>
              </div>
              <div
                class="border border-gray-700 rounded p-4"
              >
                <div
                  class="flex justify-center items-center"
                >
                  fetch is not defined
                </div>
              </div>
              <div
                class="border border-gray-700 rounded p-4"
              >
                <div
                  class="flex justify-center items-center"
                >
                  fetch is not defined
                </div>
              </div>
              <div
                class="border border-gray-700 rounded p-4"
              >
                <div
                  class="flex justify-center items-center"
                >
                  fetch is not defined
                </div>
              </div>
              <div
                class="border border-gray-700 rounded p-4"
              >
                <div
                  class="flex justify-center items-center"
                >
                  fetch is not defined
                </div>
              </div>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="border-t border-gray-700"
      >
        <div
          class="container mx-auto px-8 py-4"
        >
          <p
            class="uppercase"
          >
            Â© 
            2023
             Weather
          </p>
        </div>
      </footer>
    </div>
  </div>
</div>
`;
